### PR TITLE
Refactor the function PublicIPToString to handle both IPv4 and IPv6 in similar behavior

### PR DESCRIPTION
**Pull request by zhlsunshine**
Signed-off-by: zhlsunshine <huailong.zhang@intel.com>

1. Fixed issue for `PublicIPToString` if IPv6 address has **:** as its prefix, such as **::1** and **::c0a8:0:0:0:1**
2. Added the test case for `PublicIPToString`

### CODE REVIEW

**Review by mateiidavid — APPROVED** [review]

File: pkg/addr/addr.go [hunk 1]
```diff
@@ -29,16 +30,16 @@ func PublicAddressToString(addr *l5dNetPb.TcpAddress) string {
...
-		b = make([]byte, 16)
-		binary.BigEndian.PutUint64(b[:8], ip.GetIpv6().GetFirst())
-		binary.BigEndian.PutUint64(b[8:], ip.GetIpv6().GetLast())
+		netIP = decodeIPv6ToNetIP(ip.GetIpv6())
 	} else if ip.GetIpv4() != 0 {
-		b = make([]byte, 4)
-		binary.BigEndian.PutUint32(b, ip.GetIpv4())
+		netIP = decodeIPv4ToNetIP(ip.GetIpv4())
 	}
-	return net.IP(b).String()
```
Comment (adleong) [inline pkg/addr/addr.go:41]: hi @zhlsunshine, thanks for this fix!  Could you help me understand what caused this bug in the original code?  Originally, we would simply write the high and low halves of the ipv6 address (as uint64s) into a 16 byte array.  What does a : prefix address look like in that representation and why does `net.IP(b).String()` handle it incorrectly?

In the fixed code, it seems like there are quite a few steps:
* We write the high and low halves of the ipv6 address into a 16 byte array (b)
* We create a bigInt with those bytes
* We take the bytes back out of the bigInt (ipv6Bytes)
* We create a new byte buffer (buf) and zero pad it up the ipv6 length, then write the bytes into it
* We copy the buf into a new net.IP

Are all of these steps necessary?  Some seem a bit redundant.  In particular, our byte array (b) is already the full 16 bytes to begin with.  Do the bytes in (b) simply need to be shifted in certain cases?

Comment (zhlsunshine) [reply pkg/addr/addr.go:41]: Hi @adleong, thank you for your comment. Yes, you are correct, it's redundant here. However, I am thinking that we may represent an IPv6 address into only one `big.Int` field instead separating it into 2 halves uint64 which may caused cumbersome in later.
BTW, respectfully, I want to submit a proposal and do some API change, but I do not exactly know the process. Moreover, it should not be a good title for this PR. Of course, if you do not want to modify the ipv6 API, I also can change the function `decodeIPv6ToNetIP` as below and remove the function `IntToIPv6` :

```
func decodeIPv6ToNetIP(ip *l5dNetPb.IPv6) net.IP {
        b := make([]byte, 16)
        binary.BigEndian.PutUint64(b[:8], ip.GetFirst())
        binary.BigEndian.PutUint64(b[8:], ip.GetLast())
        return net.IP(b)
}
```
Any comment?

**Comment by zhlsunshine** [comment]
Hi @adleong Any comment for this PR?

**Comment by adleong** [comment]
Hi @zhlsunshine.  I think many of the questions in my comment above still stand.  Can you explain more about the bug that this fixes and explain what the purpose of each step is here?

**Comment by zhlsunshine** [comment]
> Hi @zhlsunshine. I think many of the questions in my comment above still stand. Can you explain more about the bug that this fixes and explain what the purpose of each step is here?

Hi @adleong, thank you for your reply, 
Firstly, I think it‘s okay for INT-format IP address converts into net.IP-format and reverse the process now in `addr.go`, because there some functions for this, such as: `IntToIPv4`, `IntToIPv6` and `IPToInt`
Secondly, for IPv6 struct as below:
```
type IPv6 struct {
	......
	First uint64 `protobuf:"fixed64,1,opt,name=first,proto3" json:"first,omitempty"` // hextets 1-4
	Last  uint64 `protobuf:"fixed64,2,opt,name=last,proto3" json:"last,omitempty"`   // hextets 5-8
}
```
I think IPv6 can be defined as following
```
type IPv6 struct {
	......
	IPv6 big.Int`protobuf:"[]byte,1,opt,name=ipv6,proto3" json:"ipv6,omitempty"` // hextets 1-4
}
```

then, specially for `IntToIPv6`, it can be okay for IPv6 INT-format to net.IP

so does it make sense?

**Comment by adleong** [comment]
I don't think that answers my questions.  I don't think there's a compelling reason to change the API.

**Comment by zhlsunshine** [comment]
> I don't think that answers my questions. I don't think there's a compelling reason to change the API.

Hi @adleong, for your question: if using current IPv6 struct, it's enough to use `binary.BigEndian.PutUint64(b[:8], ip.GetFirst())` and `binary.BigEndian.PutUint64(b[8:], ip.GetLast())` for IPv6 conversion from its struct to net.IP and unnecessary to call `IntToIPv6`. 
However, the purpose of  `IntToIPv6` is to use for handling the conversion if changing the IPv6 struct to `big.Int`.

Sure, I will respect your view if you do not want to change the current IPv6 struct. As I mentioned above, if so, I will  change the function `decodeIPv6ToNetIP` as below and remove the function `IntToIPv6` :

```
func decodeIPv6ToNetIP(ip *l5dNetPb.IPv6) net.IP {
        b := make([]byte, 16)
        binary.BigEndian.PutUint64(b[:8], ip.GetFirst())
        binary.BigEndian.PutUint64(b[8:], ip.GetLast())
        return net.IP(b)
}
```

**Comment by adleong** [comment]
Hi @zhlsunshine.  It seems like this

```
	if netIP == nil {
		return ""
	}
```

is the only change necessary to handle zero-length ip addresses.  The rest of this refactor isn't necessary and I don't think it improves readability of the code.

**Comment by zhlsunshine** [comment]
> Hi @zhlsunshine. It seems like this
> 
> ```
> 	if netIP == nil {
> 		return ""
> 	}
> ```
> 
> is the only change necessary to handle zero-length ip addresses. The rest of this refactor isn't necessary and I don't think it improves readability of the code.

Hi @adleong, yes, it might be the only necessary change from the functionality view. However, the refactor which creates a new function `decodeIPv6ToNetIP` can also make the code hierarchy more clear in `PublicIPToString` which includes handling both for IPv6 and IPv4. Does it make sense?

**Comment by adleong** [comment]
I don't think there's much to be gained from moving 3 lines of code behind a new function in this case.  Having this code inline makes the behavior more clear, in my opinion.  Also, when possible, I prefer to separate functional changes and stylistic refactors into separate pull requests.

**Comment by zhlsunshine** [comment]
Hi @adleong,  as you required,  I re-submit again to remove `decodeIPv6ToNetIP`

**Comment by adleong** [comment]
Hi @zhlsunshine, it looks like something has gone wrong with your merge.  Make sure you've pulled and merged the latest main.

**Comment by zhlsunshine** [comment]
Hi @mateiidavid & @adleong,  please let me know what else should I do to merge this PR, thanks! ^_^

**Review by adleong — APPROVED** [review]

**Comment by zhlsunshine** [comment]
Hi @adleong, I rebased my local code and re-submitted again, thanks!

**Comment by zhlsunshine** [comment]
Hi @mateiidavid & @adleong, is there anything I need to do to merge this PR? Thanks!

**Review by mateiidavid — APPROVED** [review]
Thank you for all of your work on this!

**Pull request merged**